File: src/commonMain/kotlin/com/monta/slack/notifier/SlackClient.kt
package com.monta.slack.notifier

import com.monta.slack.notifier.model.GithubEvent
import com.monta.slack.notifier.model.JobStatus
import com.monta.slack.notifier.model.JobType
import com.monta.slack.notifier.model.SlackBlock
import com.monta.slack.notifier.model.SlackMessage
import com.monta.slack.notifier.util.buildTitle
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

class SlackClient(
    private val serviceName: String?,
    private val serviceEmoji: String?,
    private val slackChannelId: String,
    private val appendAttachments: Boolean,
    private val slackHttpClient: SlackHttpClient,
) {
    suspend fun create(
        githubEvent: GithubEvent,
        jobType: JobType,
        jobStatus: JobStatus,
    ): String {
        val response = slackHttpClient.makeSlackRequest(
            url = "https://slack.com/api/chat.postMessage",
            message = generateMessageFromGithubEvent(
                githubEvent = githubEvent,
                jobType = jobType,
                jobStatus = jobStatus
            )
        )

        return requireNotNull(response?.ts)
    }

    suspend fun update(
        messageId: String,
        githubEvent: GithubEvent,
        jobType: JobType,
        jobStatus: JobStatus,
    ): String {
        val previousMessage = slackHttpClient.getSlackMessageById(messageId)

        val response = slackHttpClient.makeSlackRequest(
            url = "https://slack.com/api/chat.update",
            message = generateMessageFromGithubEvent(
                githubEvent = githubEvent,
                jobType = jobType,
                jobStatus = jobStatus,
                messageId = messageId,
                previousAttachments = previousMessage?.messages?.firstOrNull()?.attachments
            )
        )

        return requireNotNull(response?.ts)
    }

    private fun generateSlackMessageFromEvent(
        githubEvent: GithubEvent,
        serviceName: String?,
        serviceEmoji: String?,
        slackChannelId: String,
        messageId: String?,
        attachments: List<SlackMessage.Attachment>?,
    ): SlackMessage {
        val title = buildTitle(githubEvent.repository, githubEvent.workflow, serviceName, serviceEmoji)

        return SlackMessage(
            channel = slackChannelId,
            ts = messageId,
            text = title,
            blocks = listOf(
                SlackBlock(
                    type = "header",
                    text = SlackBlock.Text(
                        type = "plain_text",
                        text = title
                    )
                ),
                SlackBlock(
                    type = "divider"
                ),
                SlackBlock(
                    type = "section",
                    fields = listOf(
                        SlackBlock.Text(
                            type = "mrkdwn",
                            text = " \n*Branch:*\n${githubEvent.refName}"
                        ),
                        SlackBlock.Text(
                            type = "mrkdwn",
                            text = " \n*Run:*\n<${githubEvent.getRunUrl()}|${githubEvent.runId}>"
                        ),
                        SlackBlock.Text(
                            type = "mrkdwn",
                            text = " \n*Committer:*\n${githubEvent.displayName}"
                        ),
                        SlackBlock.Text(
                            type = "mrkdwn",
                            text = " \n*Message:*\n<${githubEvent.getChangeUrl()}|${githubEvent.getChangeMessage()}>"
                        ),
                        SlackBlock.Text(
                            type = "mrkdwn",
                            text = " \n*Change:*\n<${githubEvent.getChangeUrl()}|${githubEvent.getChangeIdentifier()}>"
                        )
                    )
                ),
                SlackBlock(
                    type = "divider"
                )
            ),
            attachments = attachments
        )
    }

    private fun generateMessageFromGithubEvent(
        githubEvent: GithubEvent,
        jobType: JobType,
        jobStatus: JobStatus,
        messageId: String? = null,
        previousAttachments: List<SlackMessage.Attachment>? = null,
    ): SlackMessage {
        val attachments = if (appendAttachments) {
            previousAttachments.orEmpty() + SlackMessage.Attachment(
                color = jobStatus.color,
                fields = listOf(
                    SlackMessage.Attachment.Field(
                        title = jobType.label + " ($LocalDateTime)",
                        short = false,
                        value = jobStatus.message
                    )
                )
            )
        } else {
            val attachments = mutableMapOf<JobType, SlackMessage.Attachment>()

            previousAttachments?.forEach { previousAttachment ->
                if (previousAttachment.jobType == null) {
                    return@forEach
                }
                attachments[previousAttachment.jobType] = previousAttachment
            }

            attachments[jobType] = SlackMessage.Attachment(
                color = jobStatus.color,
                fields = listOf(
                    SlackMessage.Attachment.Field(
                        title = jobType.label,
                        short = false,
                        value = jobStatus.message
                    )
                )
            )

            attachments.values.toList()
        }

        return generateSlackMessageFromEvent(
            githubEvent = githubEvent,
            serviceName = serviceName,
            serviceEmoji = serviceEmoji,
            slackChannelId = slackChannelId,
            messageId = messageId,
            attachments = attachments
        )
    }


    @Serializable
    data class Response(
        @SerialName("ok")
        val ok: Boolean, // true
        @SerialName("channel")
        val channel: String, // C024BE91L
        @SerialName("ts")
        val ts: String, // 1401383885.000061
    )

    @Serializable
    data class MessageResponse(
        @SerialName("ok")
        val ok: Boolean, // true
        @SerialName("messages")
        val messages: List<SlackMessage>,
    )
}

File: src/services/slack-client.ts
import { WebClient } from '@slack/web-api';
import { GitHubEvent, JobType, JobStatus, SlackMessage, SlackBlock, SlackAttachment } from '../models/types';

export class SlackClient {
  private client: WebClient;
  private serviceName?: string;
  private serviceEmoji?: string;
  private slackChannelId: string;

  constructor(config: {
    serviceName?: string;
    serviceEmoji?: string;
    slackToken: string;
    slackChannelId: string;
  }) {
    this.client = new WebClient(config.slackToken);
    this.serviceName = config.serviceName;
    this.serviceEmoji = config.serviceEmoji;
    this.slackChannelId = config.slackChannelId;
  }

  async create(
    githubEvent: GitHubEvent,
    jobType: JobType,
    jobStatus: JobStatus
  ): Promise<string> {
    const message = this.generateMessageFromGithubEvent(
      githubEvent,
      jobType,
      jobStatus
    );

    const response = await this.client.chat.postMessage({
      channel: this.slackChannelId,
      text: message.text,
      blocks: message.blocks,
      attachments: message.attachments
    });

    if (!response.ts) {
      throw new Error('Failed to create Slack message - no timestamp returned');
    }

    return response.ts;
  }

  async update(
    messageId: string,
    githubEvent: GitHubEvent,
    jobType: JobType,
    jobStatus: JobStatus
  ): Promise<string> {
    // Get previous message to preserve existing attachments
    const previousMessage = await this.getSlackMessageById(messageId);
    
    const message = this.generateMessageFromGithubEvent(
      githubEvent,
      jobType,
      jobStatus,
      messageId,
      previousMessage?.attachments
    );

    const response = await this.client.chat.update({
      channel: this.slackChannelId,
      ts: messageId,
      text: message.text,
      blocks: message.blocks,
      attachments: message.attachments
    });

    if (!response.ts) {
      throw new Error('Failed to update Slack message - no timestamp returned');
    }

    return response.ts;
  }

  private async getSlackMessageById(messageId: string): Promise<any> {
    try {
      const response = await this.client.conversations.history({
        channel: this.slackChannelId,
        oldest: messageId,
        inclusive: true,
        limit: 1
      });

      return response.messages?.[0];
    } catch (error) {
      console.error('Failed to get Slack message:', error);
      return null;
    }
  }

  private generateMessageFromGithubEvent(
    githubEvent: GitHubEvent,
    jobType: JobType,
    jobStatus: JobStatus,
    messageId?: string,
    previousAttachments?: any[]
  ): SlackMessage {
    const attachments = new Map<JobType, SlackAttachment>();

    // Preserve previous attachments
    if (previousAttachments) {
      for (const attachment of previousAttachments) {
        if (attachment.jobType) {
          attachments.set(attachment.jobType, attachment);
        }
      }
    }

    // Add/update current job status
    attachments.set(jobType, {
      color: JobStatus.getColor(jobStatus),
      jobType,
      fields: [
        {
          title: JobType.getLabel(jobType),
          value: JobStatus.getMessage(jobStatus),
          short: false
        }
      ]
    });

    return this.generateSlackMessageFromEvent(
      githubEvent,
      messageId,
      Array.from(attachments.values())
    );
  }

  private generateSlackMessageFromEvent(
    githubEvent: GitHubEvent,
    messageId?: string,
    attachments?: SlackAttachment[]
  ): SlackMessage {
    const title = this.buildTitle(githubEvent);

    const blocks: SlackBlock[] = [
      {
        type: 'header',
        text: {
          type: 'plain_text',
          text: title
        }
      },
      {
        type: 'divider'
      },
      {
        type: 'section',
        fields: [
          {
            type: 'mrkdwn',
            text: ` \n*Branch:*\n${githubEvent.refName}`
          },
          {
            type: 'mrkdwn',
            text: ` \n*Run:*\n<${this.getRunUrl(githubEvent)}|${githubEvent.runId}>`
          },
          {
            type: 'mrkdwn',
            text: ` \n*Committer:*\n${githubEvent.displayName || 'Unknown'}`
          },
          {
            type: 'mrkdwn',
            text: ` \n*Message:*\n<${this.getChangeUrl(githubEvent)}|${this.getChangeMessage(githubEvent)}>`
          },
          {
            type: 'mrkdwn',
            text: ` \n*Change:*\n<${this.getChangeUrl(githubEvent)}|${this.getChangeIdentifier(githubEvent)}>`
          }
        ]
      },
      {
        type: 'divider'
      }
    ];

    return {
      channel: this.slackChannelId,
      ts: messageId,
      text: title,
      blocks,
      attachments
    };
  }

  private buildTitle(githubEvent: GitHubEvent): string {
    const parts = [];
    
    if (this.serviceEmoji) {
      parts.push(this.serviceEmoji);
    }
    
    parts.push(githubEvent.repository);
    
    if (githubEvent.workflow) {
      parts.push(githubEvent.workflow);
    }
    
    if (this.serviceName) {
      parts.push(this.serviceName);
    }

    return parts.join(' ');
  }

  private getRunUrl(githubEvent: GitHubEvent): string {
    return `https://github.com/${githubEvent.repository}/actions/runs/${githubEvent.runId}`;
  }

  private getChangeIdentifier(githubEvent: GitHubEvent): string {
    if (githubEvent.commitSHA) {
      return githubEvent.commitSHA;
    } else if (githubEvent.prUrl) {
      return this.getPRIdentifier(githubEvent.prUrl) || 'PR';
    }
    return 'Unknown';
  }

  private getChangeUrl(githubEvent: GitHubEvent): string {
    if (githubEvent.commitSHA) {
      return `https://github.com/${githubEvent.repository}/commit/${githubEvent.commitSHA}`;
    } else if (githubEvent.prUrl) {
      return githubEvent.prUrl;
    }
    return `https://github.com/${githubEvent.repository}/`;
  }

  private getChangeMessage(githubEvent: GitHubEvent): string {
    return githubEvent.message
      ?.replace(/\n/g, ' ')
      ?.replace(/\r/g, ' ')
      ?.replace(/</g, '')
      ?.replace(/>/g, '')
      ?.substring(0, 120) || 'No message';
  }

  private getPRIdentifier(url: string): string | null {
    const regex = /pull\/\d+/;
    const match = url.match(regex);
    return match ? match[0] : null;
  }
}